### ISSUE TITLE
pureFunc is completely broken by the static argument transformation

### ISSUE THREAD

**Issue by treeowl** [bos/criterion#138]
Currently,

```haskell
pureFunc :: (b -> c) -> (a -> b) -> a -> Benchmarkable
pureFunc reduce f0 x0 = Benchmarkable $ go f0 x0
  where go f x n
          | n <= 0    = return ()
          | otherwise = evaluate (reduce (f x)) >> go f x (n-1)
```

GHC doesn't perform the static argument transformation by default (even with `-O2`), but users can turn it on manually, using `-fstatic-argument-transformation`. Furthermore, there is some talk of either getting it into shape to be able to use it by default or finding some other way to specialize functions to static arguments. If a test is compiled with `-O2 -fstatic-argument-transformation`, that transformation completely destroys the test, calculating the result *once* and then looping the given number of times. Youch!

**Comment by RyanGlScott** [comment]
I'm not sure I understand correctly... is this a `criterion` bug, or a GHC one? Moreover, this is something which normally won't bite users unless they specifically opt in to `-fstatic-argument-transformation`, right?

**Comment by treeowl** [comment]
At the moment, this will only bite users if they opt in. So at the very least, the documentation for `criterion` needs to indicate that it is incompatible with that option. But it's possible that that will change (there has been some talk about trying to find a way to specialize more aggressively to non-dictionary arguments, whether through this transformation or otherwise). To the extent you want to consider this a bug, it is a bug in `criterion`: the transformation is perfectly legitimate. I think the most obvious way to fix this is using some `NOINLINE` pragmas on the inliners of this function. Is there a reason they're marked `INLINE`? If so, that should be explained in some comments at least.

**Comment by RyanGlScott** [comment]
My (completely uneducated) guess is that it was marked `INLINE` at one point in the pursuit of making the code in `criterion` Go Really Fast™, as is usually the case with most libraries that are covered with these sorts of pragmas. I doubt there was a specific reason to inline `pureFunc` in particular.

If you could show a `criterion` program and compare the times with and without `-fstatic-argument-transformation`, and perhaps a patch that adds the necessary `NOINLINE` pragmas to make the times the same, that would be a wonderful start towards fixing this issue (or perhaps more accurately, preventing it from being an issue in the future).

**Comment by RyanGlScott** [comment]
I tried to reproduce this issue with a simple `criterion` program:

```haskell
module Main (main) where

import Criterion.Main

bigOlString :: String
bigOlString = concat $ replicate 10000 "fdsafdsfdsafdsafsd"
{-# NOINLINE bigOlString #-}

main :: IO ()
main = defaultMain [ bench "Test" $ nf (\(~x) -> bigOlString ++ x) "fdsafdsafdsa" ]
```

Where `nf` inlines `pureFunc` internally. I ran this against two versions of `criterion`: one build with and without `-fstatic-argument-transformation` enabled.

With `-fstatic-argument-transformation` enabled:

```
benchmarking Test
time                 1.630 ms   (1.613 ms .. 1.647 ms)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 1.658 ms   (1.654 ms .. 1.664 ms)
std dev              15.69 μs   (11.75 μs .. 19.79 μs)
```

Without `-fstatic-argument-transformation` enabled:

```
benchmarking Test
time                 1.640 ms   (1.635 ms .. 1.643 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 1.664 ms   (1.658 ms .. 1.670 ms)
std dev              19.09 μs   (17.04 μs .. 20.29 μs)
```

So either I did something wrong, or there's not actually a difference from using `-fstatic-argument-transformation`.

**Comment by mpickering** [comment]
Here's an example I gave Ryan a few months ago but it never made it to the ticket.

```
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE FlexibleContexts #-}

import Control.Monad
import Criterion.Main
import System.Random

import Control.Monad.State.Strict
import Control.Monad.Reader


test1mtl :: MonadState Int m => Int -> m Int
test1mtl n = foldM f 1 [0..n] where
  f acc x | x `rem` 5 == 0 = do
              s <- get
              put $! (s + 1)
              pure $! max acc x
          | otherwise = pure $! max acc x


main = do

  -- Used to definitively disable bench argument inlining
  !n <- randomRIO (1000000, 1000000) :: IO Int
  !m <- randomRIO (0, 0) :: IO Int

  let runRT = (`runReaderT`  (m :: Int))
  let runR  = (`runReader`   (m :: Int))
  let runST = (`runStateT`   (m :: Int))

  defaultMain [
    bgroup "group" [
      bgroup "group" [
        bench "RRRRS" $ whnf (runR . runRT . runRT . runRT . runST . test1mtl) n
        ]

]]
```

**Comment by RyanGlScott** [comment]
Thanks for the example, @mpickering. This example definitely shows off the ill effects of `-fstatic-argument-transformation` on `criterion`. Here's what you get when you compile the program without any optimizations:

```
benchmarking group/group/RRRRS
time                 396.8 ms   (391.7 ms .. 400.0 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 395.8 ms   (394.4 ms .. 396.5 ms)
std dev              1.183 ms   (0.0 s .. 1.269 ms)
variance introduced by outliers: 19% (moderately inflated)
```

With `-O2`:

```
benchmarking group/group/RRRRS
time                 9.030 ms   (9.024 ms .. 9.036 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 9.026 ms   (9.020 ms .. 9.033 ms)
std dev              18.06 μs   (13.30 μs .. 25.34 μs)
```

And finally, with `-fstatic-argument-transformation`:

```
benchmarking group/group/RRRRS
time                 1.591 ns   (1.588 ns .. 1.596 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 1.586 ns   (1.584 ns .. 1.590 ns)
std dev              9.043 ps   (3.769 ps .. 14.98 ps)
```

Yuck, `1.591 ns `! That's almost certainly baloney. The question now becomes how to address this problem. I attempted the advice in https://github.com/bos/criterion/issues/138#issuecomment-301271408 of marking `whnf` and `pureFunc` as `NOINLINE`, but even that did not guard them from the ill effects of `-fstatic-argument-transformation`.

One thing that _did_ work was marking the `go` function (local to `pureFunc`) as `NOINLINE`—after that change, the programs generated the same times regardless of whether the program (or `criterion` itself) was compiled with `-fstatic-argument-transformation`. I can't help but wonder if this is overkill, though. `go` itself is recursive, so would there be harm in marking it as `NOINLINE` always?

**Comment by patrickdoc** [comment]
This should be fixed by #184.

No optimization:

```
Criterion 1.3
benchmarking group/group/RRRRS
time                 794.1 ms   (786.5 ms .. 799.8 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 793.8 ms   (793.0 ms .. 794.6 ms)
std dev              1.280 ms   (0.0 s .. 1.297 ms)

HEAD
benchmarking group/group/RRRRS
time                 812.8 ms   (792.7 ms .. 836.2 ms)
                     1.000 R²   (1.000 R² .. NaN R²)
mean                 804.8 ms   (798.9 ms .. 809.4 ms)
std dev              5.777 ms   (2.826 ms .. 7.402 ms)
```

With `-O2`:

```
Criterion 1.3
benchmarking group/group/RRRRS
time                 6.885 ms   (6.848 ms .. 6.947 ms)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 6.918 ms   (6.892 ms .. 6.963 ms)
std dev              101.8 μs   (49.28 μs .. 161.2 μs)

HEAD
benchmarking group/group/RRRRS
time                 6.673 ms   (6.656 ms .. 6.689 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 6.672 ms   (6.662 ms .. 6.686 ms)
std dev              34.84 μs   (26.86 μs .. 49.72 μs)
```

With `-fstatic-argument-transformation`:

```
Criterion 1.3
benchmarking group/group/RRRRS
time                 2.742 ns   (2.732 ns .. 2.753 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 2.730 ns   (2.722 ns .. 2.744 ns)
std dev              35.52 ps   (24.80 ps .. 56.57 ps)

HEAD
benchmarking group/group/RRRRS
time                 7.452 ms   (7.420 ms .. 7.488 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 7.475 ms   (7.458 ms .. 7.493 ms)
std dev              48.24 μs   (39.18 μs .. 64.33 μs)
```

**Comment by RyanGlScott** [comment]
Excellent! I'll close this, then—do reopen if you think something else is wrong.
